agent/php_wrapper.c

```diff
@@ -59,6 +59,74 @@ nruserfn_t* nr_php_wrap_callable(zend_function* callable,
   return wraprec;
 }
 
+nruserfn_t* nr_php_wrap_generic_callable(zval* callable,
+                                         nrspecialfn_t callback TSRMLS_DC) {
+#if ZEND_MODULE_API_NO < ZEND_7_0_X_API_NO
+  char* name = NULL;
+#else
+  zend_string* name = NULL;
+#endif
+  zend_fcall_info_cache fcc;
+  zend_fcall_info fci;
+  nruserfn_t* wraprec = NULL;
+  /* not calling nr_zend_is_callable because we want to additionally populate name */
+  if (zend_is_callable(callable, 0, &name TSRMLS_CC)) {
+    /* see php source code's zend_is_callable_at_frame function to see from
+     * where these switch cases are derived */
+#if ZEND_MODULE_API_NO >= ZEND_7_0_X_API_NO
+again:
+#endif
+    switch (Z_TYPE_P(callable)) {
+      /* wrapping a string name of a callable */
+      case IS_STRING:
+#if ZEND_MODULE_API_NO < ZEND_7_0_X_API_NO
+        wraprec = nr_php_wrap_user_function(name, nr_strlen(name), callback TSRMLS_CC);
+#else
+        wraprec = nr_php_wrap_user_function(ZEND_STRING_VALUE(name),
+                                            ZEND_STRING_LEN(name),
+                                            callback TSRMLS_CC);
+#endif
+        /* It is a possibility that the wraprec returned here is a previously-instrumented,
+         * non-transient wraprec. Setting it to transient here should have no impact on
+         * functionality (as we will just remake it) and negligible performance impact. */
+        wraprec->is_transient = 1;
+        return wraprec;
+      /* wrapping an array where [0] is an object and [1] is the method to invoke
+       * the previous zend_is_callable has created the commbined object::method
+       * name for us to wrap */
+      case IS_ARRAY:
+#if ZEND_MODULE_API_NO < ZEND_7_0_X_API_NO
+        wraprec = nr_php_wrap_user_function(name, nr_strlen(name), callback TSRMLS_CC);
+#else
+        wraprec = nr_php_wrap_user_function(ZEND_STRING_VALUE(name),
+                                            ZEND_STRING_LEN(name),
+                                            callback TSRMLS_CC);
+#endif
+        wraprec->is_transient = 1;
+        return wraprec;
+      /* wrapping a closure. Need to initialize fcall info in order to wrap the
+       * underlying zend_function object */
+      case IS_OBJECT:
+        if (SUCCESS == zend_fcall_info_init(callable, 0, &fci, &fcc,
+                                            NULL, NULL TSRMLS_CC)) {
+          /* nr_php_wrao_callable already sets the is_transient flag for us */
+          return nr_php_wrap_callable(fcc.function_handler, callback TSRMLS_CC);
+        }
+      break; // Make fallthrough check happy
```

Comment:
It's not only fall through - `break` happens when `zend_fcall_info_init` failed. Maybe instead of `break`, a debug message and explicit `return NULL` would be better here?

Reply:
I prefer the break to keeps the failed-wrap log message in a single place, but I added additionally logging before the break here [f4fdb93](https://github.com/newrelic/newrelic-php-agent/pull/665/commits/f4fdb933cd347359eac847a3168d16ad3329460d)